controller_manager/src/controller_manager.cpp

```diff
@@ -62,33 +63,66 @@ ControllerManager::ControllerManager(
+    services_callback_group_);
   list_controller_types_service_ =
     create_service<controller_manager_msgs::srv::ListControllerTypes>(
-    "~/list_controller_types", std::bind(
-      &ControllerManager::list_controller_types_srv_cb, this, _1,
-      _2));
+    "~/list_controller_types",
+    std::bind(&ControllerManager::list_controller_types_srv_cb, this, _1, _2),
+    rmw_qos_profile_services_default,
+    services_callback_group_);
   load_controller_service_ = create_service<controller_manager_msgs::srv::LoadController>(
-    "~/load_controller", std::bind(
-      &ControllerManager::load_controller_service_cb, this, _1,
-      _2));
+    "~/load_controller",
+    std::bind(&ControllerManager::load_controller_service_cb, this, _1, _2),
+    rmw_qos_profile_services_default,
+    services_callback_group_);
   reload_controller_libraries_service_ =
     create_service<controller_manager_msgs::srv::ReloadControllerLibraries>(
-    "~/reload_controller_libraries", std::bind(
-      &ControllerManager::reload_controller_libraries_service_cb, this, _1,
-      _2));
+    "~/reload_controller_libraries",
+    std::bind(&ControllerManager::reload_controller_libraries_service_cb, this, _1, _2),
+    rmw_qos_profile_services_default,
+    services_callback_group_);
   switch_controller_service_ = create_service<controller_manager_msgs::srv::SwitchController>(
-    "~/switch_controller", std::bind(
-      &ControllerManager::switch_controller_service_cb, this, _1,
-      _2));
+    "~/switch_controller",
+    std::bind(&ControllerManager::switch_controller_service_cb, this, _1, _2),
+    rmw_qos_profile_services_default,
+    services_callback_group_);
   unload_controller_service_ = create_service<controller_manager_msgs::srv::UnloadController>(
-    "~/unload_controller", std::bind(
-      &ControllerManager::unload_controller_service_cb, this, _1,
-      _2));
+    "~/unload_controller",
+    std::bind(&ControllerManager::unload_controller_service_cb, this, _1, _2),
+    rmw_qos_profile_services_default,
+    services_callback_group_);
+
+  // TODO(all): Should we declare paramters? #168 - for now yes because of the tests
+  declare_parameter("robot_description", "");
+  // load robot_description parameter
+  std::string robot_description;
+  if (!get_parameter("robot_description", robot_description)) {
+    throw std::runtime_error("No robot_description parameter found");
+  }
+
+  // Declare default controller manager rate of 100Hz
+  declare_parameter("update_time_ms", 10);
+  // load controller_manager update time parameter
+  int update_time_ms;
+  if (!get_parameter("update_time_ms", update_time_ms)) {
+    throw std::runtime_error("update_time parameter not existing or empty");
+  }
+  RCLCPP_INFO(get_logger(), "update time is %.3f ms", update_time_ms);
+
+  timer_ = create_wall_timer(
+    std::chrono::milliseconds(update_time_ms),
+    std::bind(&ControllerManager::update, this),
+    realtime_callback_group_);
```

Comment:
I'm not to keen on having this update timer here, sometimes we want to trigger the updates manually from whoever owns the controller_manager.

For instance we might want to do some sequential operation in a specific order, read HW, update controller_manager, write HW.
I don't know if passing a time of 0 would disable the timer, but still feels hacky.

In my mind, the update() method is called externally (not from within the class), it does its work and executes the  `realtime_callback_group_` pending work.

Reply:
Should I add than timer in the node?

Update function should actually call "read() --> update_controllers() --> write()". I don't see why one would want to/ have to call this externally?

Especially because With new architecture the HW is dynamically loaded and manager by this node/controller manager

Reply:
Sorry I was not up to date with the new architecture, I was still thinking on the previous PR which had someone run the HW and the controller manager.

We do have a use case where we do:
write()
update_controllers() / read() simultaneously (in separate threads)

With some synchronization steps, but there are things being done in parallel to achieve 2KHz rate. There is a 1 cycle delay between update and write but it's the best we can do.

By this, I mean that we'd love that this behavior could be modified, by disabling the timer or maybe move some code around later to virtual functions that we can override.

Reply:
I agree with @v-lopez: `read() .. update_controllers() .. write()` is *not* the only way in which these calls can be scheduled, and the component making assumptions is not necessarily a good idea.

Being able to change this easily would be a good thing.

One example of an alternative (which is quite common actually): `write() .. read() .. update_controllers()`. This way you avoid jitter on the `write()`s (due to `update_controllers()` not always completing in the same amount of time).

If a re-usable loop or timer (callback) is really desired, perhaps introduce an `LoopPolicy` object (or something) which users can pass to the node and which it uses to delegate responsibility over the control flow of this particular part to users. A default implementation could do what @destogl suggests. Other users could supply their own version.

Reply:
We discussed yesterday to go for simple solution.

That is to move the parameter `update_time_ms` to the default node.